lib/hsa/mcwamp_hsa.cpp

```diff
@@ -5258,42 +5114,21 @@ void HSAOp::registerSignalCallback() {
 // Copy mode will be set later on.
 // HSA signals would be waited in HSA_WAIT_STATE_ACTIVE by default for HSACopy instances
 HSACopy::HSACopy(Kalmar::KalmarQueue *queue, const void* src_, void* dst_, size_t sizeBytes_) : HSAOp(hc::HSA_OP_ID_COPY, queue, Kalmar::hcCommandInvalid),
-    isSubmitted(false), isAsync(false), isSingleStepCopy(false), isPeerToPeer(false), future(nullptr), depAsyncOp(nullptr), copyDevice(nullptr), waitMode(HSA_WAIT_STATE_ACTIVE),
+    isAsync(false), isSingleStepCopy(false), isPeerToPeer(false), depAsyncOp(nullptr), copyDevice(nullptr), waitMode(HSA_WAIT_STATE_ACTIVE),
     src(src_), dst(dst_),
     sizeBytes(sizeBytes_)
 {
-
-
     apiStartTick = Kalmar::ctx()->getSystemTicks();
 }
 
 // wait for the async copy to complete
-inline hsa_status_t
-HSACopy::waitComplete() {
-    _wait_complete_status = HSA_STATUS_SUCCESS;
-    if (!isSubmitted)  {
-        _wait_complete_status = HSA_STATUS_ERROR_INVALID_ARGUMENT;
-        return _wait_complete_status;
-    }
-
-    // Wait on completion signal until the async copy is finishedS
-    if (DBFLAG(DB_WAIT)) {
-        hsa_signal_value_t v = -1000;
-        if (_signal.handle) {
-            hsa_signal_load_scacquire(_signal);
-        }
-        DBOUT(DB_WAIT, "  wait for copy op#" << getSeqNum() << " completion with wait flag: " << waitMode << "signal="<< std::hex  << _signal.handle << std::dec <<" currentVal=" << v << "...\n");
-    }
-
+inline void
+HSACopy::wait() {
     // Wait on completion signal until the async copy is finished
-    hsa_signal_wait_scacquire(_signal, HSA_SIGNAL_CONDITION_LT, 1, UINT64_MAX, waitMode);
-
-    isSubmitted = false;
-
-    // clear reference counts for dependent ops.
-    depAsyncOp = nullptr;
-
-    return _wait_complete_status;
+    if (_signal.handle) {
+        hsa_signal_wait_scacquire(_signal, HSA_SIGNAL_CONDITION_LT, 1, UINT64_MAX, waitMode);
+        dispose();
```

Comment:
I think `dispose()` should be called at the end of this function regardless of whether `_signal.handle` is valid.  Then in `~HSACopy()`,  only `wait()` would be needed and `dispose()` becomes redundant.

Reply:
That might be true in this case of `HSACopy`, but isn't so for `HSADispatch` (see comment below).  I agree that in some cases the call to `dispose()` becomes redundant, but the `atomic_flag` at the top of each `dispose()` implementation should be quite cheap.